internal/ingress/controller/controller_test.go

```diff
@@ -20,13 +20,192 @@ import (
+					Name:    "example-http-svc-canary-80",
+					Virtual: true,
+					VirtualMetadata: ingress.VirtualMetadata{
+						Weight: 20,
+					},
+				},
+			},
+			map[string]*ingress.Server{
+				"example.com": {
+					Hostname: "example.com",
+					Locations: []*ingress.Location{
+						{
+							Path:    "/",
+							Backend: "example-http-svc-80",
+						},
+					},
+				},
+			},
+			1,
+			1,
+		},
+		"merging a virtual backend matches with the correct host": {
+			&extensions.Ingress{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "example",
+				},
+				Spec: extensions.IngressSpec{
+					Rules: []extensions.IngressRule{
+						{
+							Host: "foo.bar",
+							IngressRuleValue: extensions.IngressRuleValue{
+								HTTP: &extensions.HTTPIngressRuleValue{
+									Paths: []extensions.HTTPIngressPath{
+										{
+											Path: "/",
+											Backend: extensions.IngressBackend{
+												ServiceName: "foo-http-svc-canary",
+												ServicePort: intstr.IntOrString{
+													Type:   intstr.Int,
+													IntVal: 80,
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+						{
+							Host: "example.com",
+							IngressRuleValue: extensions.IngressRuleValue{
+								HTTP: &extensions.HTTPIngressRuleValue{
+									Paths: []extensions.HTTPIngressPath{
+										{
+											Path: "/",
+											Backend: extensions.IngressBackend{
+												ServiceName: "http-svc-canary",
+												ServicePort: intstr.IntOrString{
+													Type:   intstr.Int,
+													IntVal: 80,
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			map[string]*ingress.Backend{
+				"example-foo-http-svc-80": {
+					Name:    "example-foo-http-svc-80",
+					Virtual: false,
+				},
+				"example-foo-http-svc-canary-80": {
+					Name:    "example-foo-http-svc-canary-80",
+					Virtual: true,
+					VirtualMetadata: ingress.VirtualMetadata{
+						Weight: 20,
+					},
+				},
+				"example-http-svc-80": {
+					Name:    "example-http-svc-80",
+					Virtual: false,
+				},
+				"example-http-svc-canary-80": {
+					Name:    "example-http-svc-canary-80",
+					Virtual: true,
+					VirtualMetadata: ingress.VirtualMetadata{
+						Weight: 20,
+					},
+				},
+			},
+			map[string]*ingress.Server{
+				"foo.bar": {
+					Hostname: "foo.bar",
+					Locations: []*ingress.Location{
+						{
+							Path:    "/",
+							Backend: "example-foo-http-svc-80",
+						},
+					},
+				},
+				"example.com": {
+					Hostname: "example.com",
+					Locations: []*ingress.Location{
+						{
+							Path:    "/",
+							Backend: "example-http-svc-80",
+						},
+					},
+				},
+			},
+			1,
+			1,
+		},
+	}
+
+	for title, tc := range testCases {
+		t.Run(title, func(t *testing.T) {
+			mergeVirtualBackends(tc.ingress, tc.upstreams, tc.servers)
+
+			numVirtualBackends := len(tc.upstreams["example-http-svc-80"].VirtualBackends)
```

Comment:
How does checking only the length ensure that the test `merging a virtual backend matches with the correct host` actually works as intended. Shouldn't the actual backend/location values be compared as well?

Reply:
My reasoning here was pretty much if the example.com host has 1 virtual backend, then the matching worked properly. 2 would mean that the `foo.bar` virtual went to the wrong real backend and 0 would mean that the `example.com` host went to the wrong real backend. Of course, this doesn't check for correctness entirely since they could be assigned in reverse order like you said. If you're okay with it, I'll improve this test when I improve the other test as well.